cli/medperf/entities/interface.py

```diff
@@ -1,77 +1,215 @@
 from typing import List, Dict, Union
-from abc import ABC, abstractmethod
+from abc import ABC
+import logging
+import os
+import yaml
+from medperf.exceptions import MedperfException, InvalidArgumentError
+from medperf.entities.schemas import MedperfBaseSchema
 
 
-class Entity(ABC):
-    @abstractmethod
-    def all(
-        cls, local_only: bool = False, comms_func: callable = None
-    ) -> List["Entity"]:
+class Entity(MedperfBaseSchema, ABC):
+    @staticmethod
+    def get_type():
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_storage_path():
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_comms_retriever():
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_metadata_filename():
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_comms_uploader():
+        raise NotImplementedError()
+
+    @property
+    def identifier(self):
+        return self.id or self.generated_uid
+
+    @property
+    def is_registered(self):
+        return self.id is not None
+
+    @property
+    def path(self):
+        storage_path = self.get_storage_path()
+        return os.path.join(storage_path, str(self.identifier))
+
+    @classmethod
+    def all(cls, unregistered: bool = False, filters: dict = {}) -> List["Entity"]:
         """Gets a list of all instances of the respective entity.
-        Wether the list is local or remote depends on the implementation.
+        Whether the list is local or remote depends on the implementation.
 
         Args:
-            local_only (bool, optional): Wether to retrieve only local entities. Defaults to False.
-            comms_func (callable, optional): Function to use to retrieve remote entities.
-                If not provided, will use the default entrypoint.
+            unregistered (bool, optional): Wether to retrieve only unregistered local entities. Defaults to False.
+            filters (dict, optional): key-value pairs specifying filters to apply to the list of entities.
+
 
         Returns:
             List[Entity]: a list of entities.
         """
+        logging.info(f"Retrieving all {cls.get_type()} entities")
+        if unregistered:
+            if filters:
+                raise InvalidArgumentError(
+                    "Filtering is not supported for unregistered entities"
+                )
+            return cls.__unregistered_all()
+        return cls.__remote_all(filters=filters)
+
+    @classmethod
+    def __remote_all(cls, filters: dict) -> List["Entity"]:
+        comms_fn = cls.__remote_prefilter(filters)
+        entity_meta = comms_fn()
+        entities = [cls(**meta) for meta in entity_meta]
+        return entities
```

Comment:
so we are not saving fetched entities, right?

Reply:
I mean is this conscious? Are there +/- comparing to use `__remote_get` and writing all the fetched entities?

Reply:
One thing we had in mind is that it could be an overkill to save them all locally (imagine a benchmark having 100 results for example). I don't see benefit of writing them btw.